### PR TITLE
ast: update for py310

### PR DESCRIPTION
**Pull request by hauntsaninja**

### CODE REVIEW

**Comment by hauntsaninja** [comment]
@erictraut is this a Pyright bug?
```
class X:
    a: a
```
gives me a NameError in Python 3.9, which suggests that the type here isn't self-referential.

**Comment by erictraut** [comment]
I don't think it's a bug in pyright.

When configured for 3.9, pyright emits the error `"a" is not defined`. That's equivalent to the Python NameError.

When configured for 3.10 (or with `from __future__ import annotations`), pyright emits the error `Type of "a" could not be determined because it refers to itself`.

**Comment by erictraut** [comment]
Come to think of it, with the recent decision to defer making 563 behavior default for Python 3.10, I guess this is actually a bug. It should generate the error `"a" is not defined` for Python 3.10 as well.

**Comment by hauntsaninja** [comment]
Okay, so `Type of "a" could not be determined because it refers to itself` makes sense to me for:
```
from __future__ import annotations
class X:
    a: a
```
But the issue here is more like the following, where I'd hope that Pyright looks at the outer scope and uses `int` for type of `X.a`.
```
from __future__ import annotations
a = int
class X:
    a: a
```
For another example, I'd expect Pyright's behaviour for the following to be similar for Python 3.9 (where X in the return type is int) and Python 3.10:
```
X = int
class X:
  def get() -> X: ...
```

**Comment by erictraut** [comment]
>I'd hope that Pyright looks at the outer scope

I disagree. When PEP 563 is in effect, all type annotations are assumed to be forward references, so code flow order is no longer in effect. Symbols in annotations should be resolved using the current scope first, and outer scopes should be consulted only if there's no resolution within the current scope.

**Comment by hauntsaninja** [comment]
Okay, so for the following code:
```
aaa = int
class X:
    aaa: aaa
x = X()
reveal_type(x.aaa)
```
This is perfectly valid on all Pythons, and should reveal `int` on Python 3.9, but Pyright reveals `aaa`. I think this is a clear bug and should be fixed.

```
from __future__ import annotations
aaa = int
class X:
    aaa: aaa
x = X()
reveal_type(x.aaa)
```
This one the desired behaviour is slightly less clear. Pyright gives me `Type of "a" could not be determined because it refers to itself`, which is defensible (it certainly shouldn't give `"aaa" is not defined`). But given that there's no useful self-referential interpretation here, I think it's still reasonable to look for in the outer scope for aaa (but of course, only when defining `X.aaa`), and this PR provides a real life case where doing so would be useful. Anyway, this point seems perhaps contentious, so I'll workaround it in this PR!

**Comment by erictraut** [comment]
>... and should reveal int

It is evaluating the type as `int`, but the `reveal_type` text is using the type alias name `aaa` instead if `int`. (We've received feedback from pyright users that they prefer that we use type alias names rather than the expanded form.)

As proof that it's actually evaluating the type as `int`, you can add the following lines, which type checks correctly:
```python
y: int = x.aaa
z: aaa = 0
```

**Comment by hauntsaninja** [comment]
Oops, should have caught that. Sorry for the noise :-)

File: stdlib/_ast.pyi [hunk 1]
```diff
@@ -375,3 +375,37 @@ class alias(AST):
...
+    # Without the alias, Pyright complains variables named pattern are recursively defined
+    _pattern = pattern
+    class match_case(AST):
+        pattern: _pattern
+        guard: Optional[expr]
+        body: typing.List[stmt]
+    class MatchValue(pattern):
+        value: expr
+    class MatchSingleton(pattern):
+        value: expr
```
Comment (JelleZijlstra) [inline stdlib/_ast.pyi:393]: According to the docs value is of type `constant` (similar to `ast.Constant`).

Comment (hauntsaninja) [reply stdlib/_ast.pyi]: Thanks, I changed to Optional[bool] because https://docs.python.org/3.10/library/ast.html#ast.MatchSingleton

**Comment by JelleZijlstra** [comment]
The stubtest failures are just 3.10 which we removed, so ignoring them.

**Pull request merged**